README
Monitoring-Generator-TestConfig
===============================

DESCRIPTION

This modul generates test configurations for nagios and icinga. This can be useful if you
want for doing load tests or testing addons and plugins.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires no other modules

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2009 by sven

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for Perl extension Monitoring::Generator::TestConfig.

0.52  Fri Feb 26 16:26:58 CET 2021
	update requirements

0.50  Wed Feb 24 14:04:57 CET 2021
	change skip_dependencys to skip_dependencies
	add underline to contact settings
	add options to skip hostgroups / servicegroups

0.48  Tue Feb 23 17:08:24 CET 2021
	add support to generate contacts and contactgroup

0.46  Mon Dec 26 19:57:30 CET 2016
	fix tests on later perl versions

0.44  Tue Oct  9 20:47:03 CEST 2012
	added receiver to shinken init script
	added option to fixed size hostnames / descriptions

0.42  Fri Dec 30 11:24:48 CET 2011
	preserve 'use lib' lines in recreate.pl
	made host/service check plugins epn compliant
	renamed default template to avoid naming conflicts with existing templates

0.40  Fri Jul  1 16:19:55 CEST 2011
	added host/servicecheck cmd line option to override check cmd
	installation on omd works now out of the box

0.38  Sat Apr 16 15:39:33 CEST 2011
	updated shinken init script

0.36  Sun Mar 20 12:56:50 CET 2011
	added omd layout

0.34  Wed Nov 17 14:05:42 CET 2010
	added init script for shinken layout
	added sample service dependencys

0.32  Thu Sep  2 13:26:43 CEST 2010
	added verbose output to init scripts config check
	added perfdata to service checks (thanks Jörg)

0.30  Thu Aug 19 21:01:57 CEST 2010
	added blocking type

0.28  Thu Apr  8 09:24:20 CEST 2010
	added shinken layout (thanks Gerhard)
    fixed tests for windows
    added multiline and html plugin output to random checks

0.26  Fri Jan  8 11:13:16 CET 2010
	fixed way of searching for binarys
	fixed config test when there is no binary found

0.24  Wed Jan  6 17:38:04 CET 2010
	renamed to Monitoring::Generator::TestConfig
	added switch to set layout (nagios/icinga for the moment)

0.22  Tue Jan  5 21:09:50 CET 2010
	fixed host/service config
	fixed problem with newlines in resource.cfg

0.20  Sun Dec 13 00:55:33 CET 2009
	added init script
	added prefix option

0.18  Sat Dec  5 15:33:10 CET 2009
	added support for parent router

0.16  Wed Dec  2 11:39:50 CET 2009
	added large installation tweaks for installations > 2000 services

0.15  Sun Nov 15 16:56:04 CET 2009
	- added possibility to override all host / service settings
	- rewritten host / service plugins in perl
	- moved service / hostplugin data to extra module
	- added sample host / servicegroups
	- added possibility to set types of hosts/services

0.14  Fri Nov 13 21:42:55 CET 2009
	- added possibility to override all nagios.cfg settings
	- added Test::Perl::Critic tests
	- added Test::Pod::Coverage tests

0.13  Tue Nov 10 09:35:29 CET 2009
	- fixed tests, now uses File::Which

0.12  Sat Nov  7 14:37:37 CET 2009
	- changed way of passing options

0.11  Thu Nov  5 23:57:17 CET 2009
	- fixed tests on solaris

0.10  Wed Nov  4 20:15:25 CET 2009
	- initial release, generates working nagios config

0.01  Tue Nov  3 18:45:19 2009
	- original version; created by h2xs 1.23 with options
		-A -X -n Nagios::Generator::TestConfig


examples/create_test_config.pl
#!/usr/bin/env perl

=head1 NAME

create_test_config.pl - create a test configuration

=head1 SYNOPSIS

./create_test_config.pl [ -h ] [ -v ] [ -b <monitoring binary> ] [ -p <prefix> ] [ -l layout ] <directory>

=head1 DESCRIPTION

this script generates a valid test configuration

=head1 ARGUMENTS

script has the following arguments

=over 4

=item help

    -h

print help and exit

=item verbose

    -v

verbose output

=item prefix

    -p

add this prefix to all exported hosts and services

=item binary

    nagios/icinga binary to use

will search for nagios, nagios3 and icinga in path if not set

=item layout

    use nagios or icinga

=item directory

    output directory for export

=back

=head1 EXAMPLE

./create_test_config.pl -p test1 /tmp/test-config/

=head1 AUTHOR

2009, Sven Nierlein, <nierlein@cpan.org>

=cut

use warnings;
use strict;
use Getopt::Long;
use Pod::Usage;
use lib '../lib';
use lib 'lib';
use Monitoring::Generator::TestConfig;

#########################################################################
# parse and check cmd line arguments
my ($opt_h, $opt_v, $opt_p, $opt_b, $opt_d, $opt_l);
Getopt::Long::Configure('no_ignore_case');
if(!GetOptions (
   "h"              => \$opt_h,
   "v"              => \$opt_v,
   "p=s"            => \$opt_p,
   "b=s"            => \$opt_b,
   "l=s"            => \$opt_l,
   "<>"             => \&add_dir,
)) {
    pod2usage( { -verbose => 1, -message => 'error in options' } );
    exit 3;
}

if(defined $opt_h) {
    pod2usage( { -verbose => 1 } );
    exit 3;
}
my $verbose = 0;
if(defined $opt_v) {
    $verbose = 1;
}

if(!defined $opt_d) {
    pod2usage( { -verbose => 1, -message => 'no export directory given!' } );
    exit 3;
}

$opt_p = ""       unless defined $opt_p;
$opt_l = "nagios" unless defined $opt_l;


#########################################################################
my $ngt = Monitoring::Generator::TestConfig->new(
                    'output_dir'                => $opt_d,
                    'layout'                    => $opt_l,
                    'verbose'                   => 1,
                    'overwrite_dir'             => 1,
                    'prefix'                    => $opt_p,
                    'binary'                    => $opt_b,
                    'routercount'               => 10,
                    'hostcount'                 => 100,
                    'services_per_host'         => 10,
                    'main_cfg'                  => {
                            'broker_module' => '/opt/projects/git/check_mk/livestatus/src/livestatus.o /tmp/live.sock',
                        },
                    'hostfailrate'              => 2, # percentage
                    'servicefailrate'           => 5, # percentage
                    'host_settings'             => {
                            'normal_check_interval' => 30,
                            'retry_check_interval'  => 5,
                        },
                    'service_settings'          => {
                            'normal_check_interval' => 30,
                            'retry_check_interval'  => 5,
                        },
                    'router_types'              => {
                                    'down'         => 10, # percentage
                                    'up'           => 10,
                                    'flap'         => 10,
                                    'pending'      => 10,
                                    'random'       => 60,
                        },
                    'host_types'                => {
                                    'down'         => 5, # percentage
                                    'up'           => 50,
                                    'flap'         => 5,
                                    'pending'      => 5,
                                    'random'       => 35,
                        },
                    'service_types'             => {
                                    'ok'           => 50, # percentage
                                    'warning'      => 5,
                                    'unknown'      => 5,
                                    'critical'     => 5,
                                    'pending'      => 5,
                                    'flap'         => 5,
                                    'random'       => 25,
                        },
);
$ngt->create();
#########################################################################

sub add_dir {
    my $dir = shift;
    $opt_d  = $dir;
}
